### ISSUE TITLE
ml restore speed

### ISSUE THREAD

**Issue by wpoely86** [TACC/Lmod#55]
Using restore:
```
$ time ml restore psi4-ints 
Restoring modules to user's psi4-ints

real	0m1.319s
user	0m0.852s
sys	0m0.296s
```
Using load directly
```
$ time ml load PSI/4.0b5-ictce-7.1.2-mt HDF5/1.8.13-ictce-7.1.2-serial

real	0m0.413s
user	0m0.276s
sys	0m0.080s
```
In both case, you end up with exactly the same list of loaded modules. Any reason why the restore is slower than a bunch of loads?

**Comment by rtmclay** [comment]
There is one difference is that a restore does a purge first then loads.  Please execute the following:

   $ ml -v
   $ ml purge
   $ time ml restore psi4-ints
   $ ml purge
   $ time ml load PSI/4.0b5-ictce-7.1.2-mt HDF5/1.8.13-ictce-7.1.2-serial

**Comment by wpoely86** [comment]
```
$ ml -v

Modules based on Lua: Version 5.9 (5.9-5-g3504ca2) 2015-02-10 05:36
    by Robert McLay mclay@tacc.utexas.edu

$ time ml purge

real	0m0.053s
user	0m0.032s
sys	0m0.008s
$ time ml restore psi4-ints
Restoring modules to user's psi4-ints

real	0m1.284s
user	0m0.832s
sys	0m0.296s
$ time ml purge
Lmod Warning:  The environment MODULEPATH has been changed in unexpected ways.
Lmod is unable to use given MODULEPATH. It is using:
    "/home/opt/easybuild/modules/all:".
Please use "module use ..." to change MODULEPATH instead. 

real	0m0.362s
user	0m0.220s
sys	0m0.096s
$ time ml load PSI/4.0b5-ictce-7.1.2-mt HDF5/1.8.13-ictce-7.1.2-serial

real	0m0.397s
user	0m0.264s
sys	0m0.084s
```
So purge + load is still faster then restore?

**Comment by rtmclay** [comment]
I'm not sure why.  I'll have to look at it.   It may be related to issue #53.  Please do the following:

    $ ml purge
    $ ml -D restore psi4-ints  2> restore.log

and email me restore.log

**Comment by rtmclay** [comment]
I looked at your restore.log and it is not a fair comparison.  Your collections loads 16 modules where as loading 4 modules is going to be slower.

**Comment by wpoely86** [comment]
Yes, but those 2 modules load other modules as dependencies. In the end, you have 16 modules in both cases?

**Comment by rtmclay** [comment]
Now I think I know what is going on.  Loading a collection does exactly that.  It loads them in the order specified.   But if inside a modulefile it loads the same module that it already loaded. it reloads them.   I just checked.  Each of the 16 modules are loaded twice.  This is not an Lmod bug but a feature. If a modulefile request a load, it loads them.  If you do a purge and then load PSI and HDF5, it probably only loads them once.

Because of the fact that your modulefiles are currently written in TCL,  I don't think that you can avoid this problem.  I suspect that the TCL modulefiles are checking to see if a module is not loaded before it tries to load it.  But because of the time when Lmod is doing the translation from TCL -> Lua, the test to see if it loaded fails because it wasn't loaded at the time the translation happens.

**Comment by wpoely86** [comment]
OK, I understand. A native question: can't you skip the whole dependencies loading and check if it's already loaded on a restore? All of that was already one on the actual load before the save. A restore should just load the modules in the correct order without worrying about deps and stuff because it already checked all of that on the original load before the save and it starts from a purge?

**Comment by rtmclay** [comment]
But all restore does is purge then load the requested modulefiles.    All these problems will go away when EB starts building Lua based modules.

**Comment by wpoely86** [comment]
OK, I will poke @boegel until they work.

**Comment by boegel** [comment]
keep an eye on https://github.com/hpcugent/easybuild-framework/pull/1060, it's close to being ready to be merged in (but it won't happen before EB v2.0)
